File: RSA Digital Signature.py
#In normal RSA, public key for encryption, private key for decryption
#For RSA Digital Signature, Encryption with Private Key, Decryption with Public Key
#public key = e
#private key = d

import hashlib

def gcd(a, b):
    if(b > a):
        a, b = b, a
    c = a % b
    if(c == 0):
        return b
    return gcd(b, c)

def pow_big(a, b, n):
    if b == 0:
        return 1
    if(b % 2 == 0):
        result = pow_big(a, b//2, n)
        return ((result % n) * (result % n)) % n
    else:
        return ((a % n) * (pow_big(a, b - 1, n) % n)) % n
    
    
    
p = 656692050181897513638241554199181923922955921760928836766304161790553989228223793461834703506872747071705167995972707253940099469869516422893633357693
q = 204616454475328391399619135615615385636808455963116802820729927402260635621645177248364272093977747839601125961863785073671961509749189348777945177811

n = p * q

e = 2
phi_n = (p - 1) * (q - 1)

while(e < phi_n):
    if gcd(e, phi_n) == 1:
        break
    else:
        e += 1


d = 1
for k in range(0, n):
    x = 1 + k * phi_n
    if(x % e == 0):
        d = x // e
        break
    

# print("K: ", k)
print("Public Key(Encryption): ", e)
print("Private Key(Decryption): ", d)
print("\n\n")



msg = 1807034343434343434
hash_object = hashlib.sha256(str(msg).encode())
hex_hash = hash_object.hexdigest()
hash_msg = int(hex_hash, 16)
print("The Message: ", msg)
# print("The Hash Message: ", hex_hash)
print("The Hash Message: ", hash_msg)
encrypted_digest = pow_big(hash_msg, d, n)
print("The Encrypted Digest: ", encrypted_digest)

#The sender now sends the original message with the encrypted has value
#The receiver now decrypts the encrypted value

print()
print()
expected_digest = pow_big(encrypted_digest, e, n)
print("The Expected(Decrypted) Digest: ", expected_digest)
actual_hash_object = hashlib.sha256(str(msg).encode())
actual_hex_hash = hash_object.hexdigest()
actual_digest = int(hex_hash, 16)
print("The Actual Digest: ", actual_digest)

print()
if(actual_digest == expected_digest):
    print("As the expected digest and actual digest are same, the signature is verified i.e. the message is okay.")

File: RSA Digital Signature.cpp
//In normal RSA, public key for encryption, private key for decryption
//For RSA Digital Signature, Encryption with Private Key, Decryption with Public Key
//public key = e
//private key = d


#include <bits/stdc++.h>
using namespace std;

long long int gcd(long long int a, long long int b)
{
    if(b > a)
        swap(a, b);
    long long int c = a % b;
    if(c == 0)
        return b;
    return gcd(b, c);
}

long long int pow_big(long long int a, long long int b, long long int n)
{
    if(b == 0)
        return 1;
    if(b % 2 == 0)
    {
        long long int result;
        result = pow_big(a, b / 2, n);
        return ((result % n) * (result % n)) % n;
    }
    else
        return ((a % n) * (pow_big(a, b - 1, n) % n)) % n;
}


int main()
{
    long long int i, a, b, c, d, n, p, q, m1, m2, e, phi_n, x, e1, e2, m_e, msg, d_e;
//    a = 12, b = 16;
//    cout<<gcd(b, a)<<endl;

    p = 7907;
    q = 7919;

    n = p * q;
    phi_n = (p - 1) * (q - 1);

    e = 2;
    while(e < phi_n)
    {
        if(gcd(e, phi_n) == 1)
            break;
        e++;
    }


    //one way of finding d
    d = 1;
    for(i = 0; i < n; i++)
    {
        x = 1 + i * phi_n;
        if(x % e == 0)
        {
            d = x / e;
            break;
        }
    }


    cout<<"The e: "<<e<<endl;
    cout<<"The n: "<<n<<endl;
    cout<<"The d: "<<d<<endl<<endl;



    msg = 1807034;
    hash<string>hash_fn;
    size_t msg_hash = hash_fn(to_string(msg)) % n;

    cout<<"The message: "<<msg<<endl<<endl;
    cout<<"The hash message: "<<msg_hash<<endl<<endl;

    m_e = pow_big(msg_hash, d, n);
    cout<<"The encrypted hash: "<<m_e<<endl<<endl;

    cout<<"In receiver side:\n-----------------"<<endl<<endl;

    size_t rcv_hash = hash_fn(to_string(msg)) % n;
    cout<<"The calculated hash of message: "<<rcv_hash<<endl<<endl;

    d_e = pow_big(m_e, e, n);
    cout<<"The decrypted hash: "<<d_e<<endl<<endl;

    if(d_e == rcv_hash)
        cout<<"VERIFIED!!!"<<endl;
    else
        cout<<"Verification failed!!!"<<endl;
    return 0;

}
